### DISCUSSION TITLE
Use Pango WrapMode in sound applet [Feature request]

### DISCUSSION THREAD

**Discussion by Hilyxx** [linuxmint/discussions#79] [Ideas]
Sound applet is great but with long titles, we lost a lot of information. Could you please add (in the future) this functionality ?

 **Here are some mockups:**
 
 Without WrapMode
![Capture d’écran du 2023-07-16 16-19-33](https://github.com/linuxmint/discussions/assets/21991448/4804d863-99f0-4d5b-8367-050bfc9fd705)

With WrapMode
![Capture d’écran du 2023-07-16 16-14-41](https://github.com/linuxmint/discussions/assets/21991448/eb3511d5-c899-4e17-8802-bbdde00e4bff)

Thanks

**Comment by claudiux** [comment]
Have you a piece of code to offer, please?

**Reply by Hilyxx** [reply]
Bonjour claudiux 
Je vous réponds en français (j'ai vu que vous l'êtes aussi) parce que l'anglais et moi..

Malheureusement je ne suis pas développeur  même si j'aurais aimé contribuer à linux mint de cette façon. 

J'ai essayé avec les très maigres connaissances que j'ai et en regardant un peu ailleurs et j'ai testé avec ça dans l'applet son:

A partir de la ligne -- 561 --

 this.titleLabel = new St.Label({text: this._title,  x_expand: true, x_align: Clutter.ActorAlign.START});

        const clutterText = this.titleLabel.get_clutter_text();
        clutterText.line_wrap = true;
        clutterText.line_wrap_mode = WrapMode.WORD_CHAR;
        clutterText.ellipsize = EllipsizeMode.NONE;

En rajoutant "const {EllipsizeMode, WrapMode} = imports.gi.Pango; " au début .

Oui le code peut paraitre affreux mais cela fonctionne. Et en améliorant le style css on a quelque chose d'assez fonctionnel.

Par contre le problème qui se pose c'est que le titre ne s'affiche pas du tout parfois comme s'il avait du mal à charger. Mais ça le fait uniquement quand le menu de l'applet n'est pas affiché. Si l'applet est affiché sur le bureau comme sur les mockups  ou qu'il est rafraichi cela fonctionne.

Voila pour la remontée d'infos. 

C'était juste une demande comme ça. J'écoute beaucoup de musique et l'applet radio sur cinnamon fonctionne de pair avec l'applet son via mpris mais beaucoup de titres sont très long et c'est dommage d'avoir la moitié du titre d'affiché à chaque fois. Et les applets radio ne prennent pas non plus en charge cette fonction donc je fais ce que je peux :) ...

**Reply by rcalixte** [reply]
@Hilyxx I modified your code and the CSS bug. The PRs if you want to test this on your machine: linuxmint/mint-themes#453, linuxmint/cinnamon#11756.

Thank you for the idea!


Le français et moi... 😅 

J'ai modifié votre code et le bug CSS. Les demandes d'extraction (toujours "pull requests"? 😅) si vous voulez tester cela sur votre machine: linuxmint/mint-themes#453, linuxmint/cinnamon#11756.

Merci pour l'idée!

**Reply by Hilyxx** [reply]
@rcalixte Thank you for your commit and your reply.

I have tested your code and issue with title is always here. Sometimes titles are full displayed correctly and sometimes no. This issue occurs when sound applet is not displayed. Tested with spotify, lollypop and radio applet.

It's not a css bug, i tried with my theme and your commit in official theme. Maybe applet is not "active" in background when is not displayed and titles fails to load with WrapMode

![Capture d’écran du 2023-07-17 00-16-35](https://github.com/linuxmint/discussions/assets/21991448/3807a62d-721e-4dad-a530-fd1aecfc6293)

Full title song is "It's a Long Way to the Top (If You Wanna Rock 'N' Roll")

**Reply by rcalixte** [reply]
That's interesting. I tested with some long names across several applications but I couldn't generate that. Let me test with that exact file to see if I can replicate.

One other difference is my system is in English. I wonder if that comes into play as well. I'll let you know what I can find! Thanks for the explanation!

**Reply by Hilyxx** [reply]
yes of course

File in zip, github doesn't want to share js files..
[applet.zip](https://github.com/linuxmint/discussions/files/12065284/applet.zip)

**Reply by rcalixte** [reply]
![2023-07-16_20-16-31](https://github.com/linuxmint/discussions/assets/10281587/f66021a1-4b95-49b1-9f08-c0a1a3764669)
![2023-07-16_20-15-36](https://github.com/linuxmint/discussions/assets/10281587/caff274e-8091-4223-a008-1e17cd49e9cf)
![2023-07-16_20-44-35](https://github.com/linuxmint/discussions/assets/10281587/c1615e19-7f49-4f94-9cb2-9caff2c805e4)


I moved the applet to the same position and played the audio file via VLC and Firefox.

I'm going to test with a different localization other than English. Can you try with a Mint-Y theme, using the change from the PR to remove the height property?

**Reply by Hilyxx** [reply]
This issue is hard to explain. I  made a video.
While track changes, sound applet and spotify are not displayed, just in background

https://github.com/linuxmint/discussions/assets/21991448/2de66f13-3bb9-47c7-8a1f-8c3bea352241

**Comment by claudiux** [comment]
Search about 'substr' in the code of this applet...

**Reply by Hilyxx** [reply]
j'ai trouvé 'substr' dans le code, je l'ai désactivé, j'ai aussi testé en augmentant la limite de caractères, le problème est toujours là. C'est vraiment quand l'applet est arrière plan.. les titres deviennent feignants.

**Comment by claudiux** [comment]
I think I have the solution... I'll post a new PR for sound150@claudiux.

**Comment by claudiux** [comment]
PR made: https://github.com/linuxmint/cinnamon-spices-applets/pull/4946 and merged.
Please try sound150@claudiux v4.1.0 instead of sound@cinnamon applet. https://cinnamon-spices.linuxmint.com/applets/view/306

**Reply by Hilyxx** [reply]
Thank you . I have tried sound150@claudiux v4.1.0 and it's almost solved. I haven't issues with invisible titles. It's a great improvement .. But sometimes word wrap isn't displayed .

**Original applet** 
![Original applet](https://github.com/linuxmint/discussions/assets/21991448/bb542994-fa5f-4b1d-818f-ad3693908acd)

**sound150@claudiux v4.1.0**
![Applet 150](https://github.com/linuxmint/discussions/assets/21991448/46e369f6-b240-4497-9c8e-bc4be2377f9b)

**Comment by claudiux** [comment]
> I haven't issues with invisible titles.

Great!

> But sometimes word wrap isn't displayed.

Could you be more specific? Give some examples?

**Reply by Hilyxx** [reply]
Je vais essayé d'être le plus précis possible meme si le bug en question me dépasse un peu ..

Comme je l'ai dit le problème de titres invisibles ou tronqués est je pense résolu grace à votre changement. 

Par contre, et c'est encore aléatoire, le retour à la ligne des titres ne s'affiche pas toujours. 

J'ai remarqué (et finalement c'était le même probleme que précédemment) que quand l'applet est affiché au premier plan (comme sur les photos) en meme temps que le changement de titre, il n'y a aucun problème. 

En revanche quand l'applet n'est pas affiché et que le titre change, le retour à la ligne ne se fait pas (ou pas tout le temps) et donc il manque les derniers mots qui devraient être affichés en dessous. Que le lecteur (spotify ou autre) soit affiché ou minimisé ne change rien.

**Comment by claudiux** [comment]
Si vous aviez des exemples de titres, en précisant comment ils apparaissent (où se font les césures, où elles devraient se faire et ne se font pas), cela nous permettrait de mieux cerner le problème. De mon côté, je n'ai rien remarqué pour l'instant.

If you had some examples of titles, specifying how they appear (where the caesura occurs, where it should occur and where it doesn't), that would help us understand the problem better. For my part, I haven't noticed anything yet.

**Reply by Hilyxx** [reply]
_**J'ai pris pour exemple le theme officiel mint y legacy avec la police de caractère ubuntu taille 10.**_ 

Dans les exemples ci-dessous les titres sont ceux de chansons diffusées via l'applet radio. (mais c'est valable pour un lecteur lambda comme spotify ou lollypop par exemple)

La première image est celle de l'applet original sans les changements effectués de votre part. On voit que les 3/4 des mots du titre ont disparus.
Word wrap fonctionne mais aléatoirement en plus du bug de mot invisible et il faut rafraichir le code de l'applet pour que le titre apparaissent normalement. ( mais le problème réapparait juste après). 
![original](https://github.com/linuxmint/discussions/assets/21991448/687c6be7-4370-4796-9a8f-2412490ca859)


La deuxième image est celle de votre applet. Même titre. Le bug des mots invisibles a disparu mais la césure ne s'est pas faite au niveau de la fin de la ligne. Donc le mot "godfather" qui n'a pas la place pour la largeur du menu n'apparait pas et est même coupé au niveau des lettres sans retour à la ligne. (ce qui ne devrait pas arriver avec WORD_CHAR normalement). Problème aléatoire avec une césure qui peu se faire ou pas  (la je ne comprends pas..). _Ici je suis allé changer la radio et le titre s'est chargé (en arrière plan) AVANT que j'appuye pour afficher le menu de Sound150. et que je regarde le titre._
![150 issue](https://github.com/linuxmint/discussions/assets/21991448/d9e1a2ae-105e-4236-abae-3b5b7fe9fc77)

La troisième est toujours votre applet mais après avoir rafraichit le code via looking glass.
![150 refresh](https://github.com/linuxmint/discussions/assets/21991448/0c93bc92-ad7f-4e80-af4b-922f9b0c805f)


**_Deuxième exemple avec un titre différent_**

La première image est celle de l'applet original (la photo parle d'elle même)
![a](https://github.com/linuxmint/discussions/assets/21991448/6e4d264c-94d4-4bac-b2f2-be9fae84c5de)


La deuxième image (votre applet) montre encore un titre qui ne revient pas à la ligne avec une césure qui ne s'est pas faite à la fin du menu. Comme précédemment avec la deuxième image, je suis allé changer la radio et le titre s'est chargé (en arrière plan) AVANT que j'appuye pour afficher le menu de Sound150 et que je regarde le titre.
![b](https://github.com/linuxmint/discussions/assets/21991448/57e285ba-4238-4629-9563-d51884ace144)


Pour la troisème image, je n'ai pas comme précédement rafraichi le code mais j'ai ouvert le menu de l'applet AVANT que le titre n'apparaisse. Et là aucun problème, la césure a bien a été faite au niveau des mots avec un retour à la ligne au niveau du mot "immer". (j'ai répété l'opération plusieurs fois et avec succès) 
![c](https://github.com/linuxmint/discussions/assets/21991448/3f93b704-25ff-43ea-a339-b7ab2d66d454)

_Je me suis demandé si ce n'était pas non plus un problème avec Unicode de ma LMDE5. Donc j'ai testé sur une nouvelle installation toute fraiche (linux mint 21.2) mais le problème est également présent._

**Comment by claudiux** [comment]
Odd. The `...` displayed do not come from the code of the sound150 applet. You did remove the sound@cinnamon applet from the panel, didn't you?

https://gitlab.gnome.org/GNOME/pango/-/issues/243

**Reply by Hilyxx** [reply]
Non désolé j'ai oublié que les applets partagaient le même code... je viens de le refaire avec les mêmes étapes que dans mon précédent message mais le résultat est le même (sans les ... )

![b2](https://github.com/linuxmint/discussions/assets/21991448/6790137d-8f9e-4c31-ba36-90de2b473980)

![a2](https://github.com/linuxmint/discussions/assets/21991448/1f84ccdb-2405-423d-be65-53a7e9a36ac2)

**Comment by claudiux** [comment]
@Hilyxx Please try to replace the line 74 of the applet.js script by this one:
`const words = text.trim().replace(/[\r\n]+/g, ' ').split(' ');`
(just adding `.trim()` after `text`).

Then, reload sound150. (The xsession applet can help you.)

**Reply by Hilyxx** [reply]
I tried with your commit but issue is always here..
![Capture d’écran du 2023-07-17 17-16-01](https://github.com/linuxmint/discussions/assets/21991448/fef445b6-8f72-46d6-940f-427d33b9f811)
![Capture d’écran du 2023-07-17 17-15-36](https://github.com/linuxmint/discussions/assets/21991448/cbe54672-5996-4705-bd7e-9f05a7a8feee)

**Comment by ghost** [comment]
@claudiux

Could this happen due to certain restricted chars not being (properly) escaped in some regular expression? Just an idea...

On another note: would there be any such fixes/improvements in your applet(s) implemented for older Cinnamon versions too, or would those be treated as obsolete? I'm asking because I plan on staying with Mint 19.2 for as long as possible, and as such would like to avoid unnecessary updates if the new code/changes wouldn't pertain to that OS version, by renaming the applet(s).

**Reply by claudiux** [reply]
I'll explore your idea, thank you.

LM 19.2 with Cinnamon 4.2? I don't know. I can try for Cinnamon 4.4 (Cinnamon 19.3) but no less.

**Reply by ghost** [reply]
Yes, it's Cinnamon 4.2.4 here.
Mint 19.3 already started downgrading/removing certain features I use daily so won't ever touch it or anything higher.
Thank you for your reply.

**EDIT:**
On my system I've duplicated the **3.4** folder renaming it as **4.2**, and added the recent changes discussed in this topic. At first glance it works in Cinnamon 4.2. Couldn't reproduce OP's issues - maybe it only happens with ID3 text...? Tried audio files with long names **and** ID3s, and also video files - which included apostrophe and parenthesis - all displayed fine.
Player used was **XPlayer 2.2.6**, system language is English.

**Comment by claudiux** [comment]
@Hilyxx I'll keep looking, but I'm afraid I can't do much better.

**Reply by Hilyxx** [reply]
yes it's an elusive bug... :( 
Thank you @claudiux for your help. I 'll keep Sound150 applet with your commits. it works fine despite the small issue with my system.

**Comment by claudiux** [comment]
It seems that version 4.1.1 of `sound150@claudiux` solves this problem. I will propose the same changes for `sound@cinnamon.org`.

Thanks for your help, @Hilyxx.

**Reply by Hilyxx** [reply]
yes,  i have no problems with version 4.1.1 now. 
Linux mint has a great team. Thank you
